source/Variant.cpp

```diff
@@ -0,0 +1,306 @@
+		bool remove = (child.Token(0) == "remove");
+		if((add || remove) && child.Size() == 1)
+		{
+			child.PrintTrace("Skipping invalid \"" + child.Token(0) + "\" tag:");
+			continue;
+		}
+		bool variant = (child.Token(add || remove) == "variant");
+
+		if(remove)
+		{
+			if(variant)
+			{
+				// If given a full definition of one of this variant's variant members, remove the variant.
+				Variant toRemove(child);
+				auto VariantToRemove = [&](const WeightedVariant &v) noexcept -> bool
+				{
+					return v.Get() == toRemove;
+				};
+
+				auto removeIt = remove_if(variants.begin(), variants.end(), VariantToRemove);
+				if(removeIt != variants.end())
+					variants.erase(removeIt, variants.end());
+				else
+					child.PrintTrace("Warning: Did not find matching variant for specified operation:");
+			}
+			else
+			{
+				// If given the name of a ship, remove all ships by that name from this variant.
+				string shipName = child.Token(1);
+				auto ShipToRemove = [&](const Ship *s) noexcept -> bool
+				{
+					return s->VariantName() == shipName;
+				};
+
+				auto removeIt = remove_if(ships.begin(), ships.end(), ShipToRemove);
+				if(removeIt != ships.end())
+					ships.erase(removeIt, ships.end());
+				else
+					child.PrintTrace("Warning: Did not find matching ship for specified operation:");
+			}
+		}
+		else
+		{
+			// If this is a subsequent call of Load(), clear the variant
+			// if we aren't adding to it.
+			if(reset && !add)
+			{
+				reset = false;
+				variants.clear();
+				ships.clear();
+			}
+
+			int n = 1;
+			int index = 1 + add;
+			if(variant)
+			{
+				string variantName;
+				if(child.Size() >= index + 1 && !child.IsNumber(index))
+				{
+					variantName = child.Token(index++);
+					if(variantName == name)
+					{
+						node.PrintTrace("Warning: A variant cannot reference itself:");
+						continue;
+					}
+				}
+
+				if(child.Size() >= index + 1 && child.Value(index) >= 1.)
+					n = child.Value(index);
+
+				// If this variant is named, then look for it in GameData.
+				// Otherwise this is a new variant definition only for this variant.
+				if(!variantName.empty())
+				{
+					variants.emplace_back(GameData::Variants().Get(variantName), n);
+					if(child.HasChildren())
+						child.PrintTrace("Warning: Skipping children of named variant in variant definition:");
+				}
+				else
+					variants.emplace_back(child, n);
+			}
+			else
+			{
+				if(child.Size() >= index + 1 && child.Value(index) >= 1.)
+					n = child.Value(index);
+				ships.insert(ships.end(), n, GameData::Ships().Get(child.Token(index - 1)));
+			}
+		}
+	}
+}
+
+
+
+void Variant::FinishLoading()
+{
+	if(name.empty())
+		return;
+	
+	// Prevent a named variant from containing itself.
+	for(auto it = variants.begin(); it != variants.end(); )
+	{
+		if(it->Get().NestedInSelf(name))
+		{
+			it = variants.eraseAt(it);
+			Files::LogError("Error: Infinite loop detected and removed in variant \"" + name + "\".");
```

Comment:
> What's stopping this from being done in load?

I was probably spooked by potential load order scenarios, but I believe that no matter the load order, any loop will eventually get caught.

Suppose variant A contains variant B, which contains variant C, which contains variant A.
If variant A loads first it would be checking if variant B contains itself somewhere, yet neither variant B nor variant C have been loaded. That's the scenario that had me making this method. But then once variant C loads, it will see that it contains itself through variant A and remove variant A, so the loop does eventually get broken.